README.md
# Endless Canyon
A 3D shooter game made as a Grade 10 computer science project.


Summative/Changelog.txt
25-04-2017
 - added "rubiks cube" 3d model to find display/graphics bugs
 - fixed point projection issues
 - added depth sorting for displaying triangles
 - added 3-axis sideways mouse rotation system
 - added wsad shift space system for movement
 - bug fixes
27-04-2017
 - added terrain generation and display algorithm
 - stopped drawing faces facing away from the camera
 - created a "face" class
 - added basic lighting
 - bug fixes
30-04-2017
 - fixed lighting bug
02-05-2017
 - added sky
 - improved randomization function
 - created "player" class
 - changed movement system to control player accelleration
 - added gravity
 - added drag
 - added angular velocity
 - fixed 3d rotation algorithm issues
03-05-2017
 - added terrain size scaling
04-05-2017
 - fixed bug where triangles with exactly 2 vertices behind the camera could not display
06-05-2017
 - fixed bug where triangles with exactly 1 vertex behind the camera could not display (took 4 hours)
 - altered terrain generation inputs
 - fixed algorithm for identifying the away-facing faces of objects
 - removed rubiks cube
07-05-2017
 - made graphics system slightly faster
 - added Aircraft class
 - added rotational accelleration
 - adjusted movement mechanics
08-05-2017
 - completed Aircraft class, now 12 numbers are required to define any aircraft.
 - adjusted movement
09-05-2017
 - adjusted movement
 - added helicopter
10-05-2017
 - fixed lag
13-05-2017
 - added bullets
 - added enemies
15-05-2017
 - added particle effects
 - fixed enemies
 - added enemy crash effects
17-05-2017
 - created a version which could be imported
18-05-2017
 - split the game into 2 files: game and menu system
19-05-2017
 - created menu tree
 - created menu tree class
20-05-2017
 - added button images
 - increased airplane speed
 - decreased mountain size
 - added crosshairs and aim circle
21-05-2017
 - created instructions image
 - enemies now spawn one by one
 - added kill count
 - holding the mouse button now shoots continuously
 - added dirt clump model for particle effects
22-05-2017
 - added sounds
 - added missiles
23-05-2017
 - fixed bug resulting in death at spawn
 - missiles now face the enemy
 - fixed missile target prioritizing
 - fixed lag spike issues (finally)
 - added missile to instructions
 - added background
 - fixed button images
 - added title


Summative/main4.py
#ENDLESS CANYON Main Code
#Isaac Liao
#version 1
#23-05-2017
import pygame, sys, time, random
from pygame.locals import*
sys.path.append(sys.path[0] + '\\modules')
import game4 as game
from math import*
from numpy import*

#screen class for menu tree
class Game_Screen():
    def __init__(self, buttons, texts):
        self.buttons = buttons
        self.texts = texts
        self.divs = [] # list of sections of screen consisting of 2 parts: colour and rect.
        game.setup(window)
    def display(self):
        global window
        for division in self.divs:
            pygame.draw.rect(window, division[0], division[1])
        for obj in self.buttons+self.texts:
            obj.display()
            
#button class for menu tree
class Button():
    def __init__(self, rect, destination, x=0, y=0, pushed_rect=None): #pushed_rect is the image of the pushed button
        self.rect = rect
        if type(rect) == pygame.Surface: #if the button is an image
            self.x = x
            self.y = y
            self.pushed_rect = pushed_rect
        self.destination = destination # where the button leads
        self.pushed = False
    def press(self, tree, current_screen):
        if self.destination == 'back':
            current_screen = go_back(tree, current_screen)
        elif type(self.destination) == Screen or type(self.destination) == Game_Screen:
            return self.destination
        return current_screen
    def display(self):
        global window
        if type(self.rect) == pygame.Rect:
            pygame.draw.rect(window, (128, 128, 128), self.rect)
        elif type(self.rect) == pygame.Surface:
            w, h = self.rect.get_width(), self.rect.get_height()
            if self.pushed and self.pushed_rect != None: window.blit(self.pushed_rect, (self.x-w/2, self.y-h/2))
            else: window.blit(self.rect, (self.x-w/2, self.y-h/2))
    def check_click(self, x, y):
        if type(self.rect) == pygame.Rect:
            return self.rect.collidepoint(x, y)
        if type(self.rect) == pygame.Surface:
            w, h = self.rect.get_width(), self.rect.get_height()
            return pygame.Rect(self.x-w/2, self.y-h/2, w, h).collidepoint(x, y)

#screen class for menu tree
class Screen():
    def __init__(self, background, buttons, texts, images=[]):
        self.buttons = buttons
        self.texts = texts
        self.background = background
        self.images = images
    def display(self):
        global window
        if type(self.background) == tuple:
            window.fill(self.colour)
        elif type(self.background) == pygame.Surface:
            window.blit(self.background, (0, 0))
        for obj in self.buttons+self.texts+self.images:
            obj.display()

#image class for menu tree
class Image():
    def __init__(self, image, x, y):
        self.image = image
        self.x = x
        self.y = y
    def display(self):
        global window
        w, h = self.image.get_width(), self.image.get_height()
        window.blit(self.image, (self.x - w/2, self.y - h/2))

#text class for menu tree
class Text():
    def __init__(self, string, size, font, colour, x, y, align = 'center'):
        self.string = string
        self.align = align
        self.x = x
        self.y = y
        self.size = size
        self.colour = colour
        self.font = font
        self.update()
    def update(self): #change properties
        lines = self.string.split('\n')
        fontobj = pygame.font.Font(self.font, int(self.size))

        #special calculation must be done because text.render() does not support newline characters
        max_w = 0
        total_h = 0
        for line in lines: #find the maximum width
            w, h = fontobj.size(line)
            max_w = max(max_w, w)
            total_h += h

        self.textobj = pygame.Surface((max_w, total_h), flags=SRCALPHA)
        
        dy = 0 #dy is the vertical displacement
        for line in lines: #add lines vertically
            w, h = fontobj.size(line)
            nextline = fontobj.render(line, False, self.colour) #create a surface with text, blit it below the previous line.
            if self.align == 'left':
                self.textobj.blit(nextline, (0, dy))
            elif self.align == 'center':
                self.textobj.blit(nextline, (max_w/2-w/2, dy))
            dy += h
        if self.align == 'left':
            self.top, self.left = int(self.x), int(self.y-dy/2)
        elif self.align == 'center':
            self.top, self.left = int(self.x-max_w/2), int(self.y-dy/2)
    def display(self):
        global window
        window.blit(self.textobj, (self.top, self.left))
        
#get_all_screens() returns all screens in the menu tree
def get_all_screens(screen):
    screen_list = [screen]
    for button in screen.buttons:
        if button.destination == None:
            continue
        if type(button.destination) == Screen or type(button.destination) == Game_Screen:
            for listed_screen in get_all_screens(button.destination):
                screen_list.append(listed_screen)
    return screen_list

#go_back() returns to the parent screen for any screen, quits if at the main menu
def go_back(tree, current_screen):
    if current_screen == tree:
        pygame.quit()
        sys.exit()
    for possible_parent in get_all_screens(tree):
        for button in possible_parent.buttons:
            if button.destination == current_screen:
                return possible_parent

#window setup
pygame.init()
width = 1000
height = 1000
background_colour = (100, 150, 255)
window = pygame.display.set_mode((width, height))

#edit the button images to remove unwanted corners
back_button = pygame.image.load('images\\back icon.jpg')
back_button.set_colorkey((255, 255, 255))
pygame.draw.polygon(back_button, (255, 255, 255), [(164, 108), (148, 108), (164, 92)]) #get rid of the white triangle in the corner
back_button.convert_alpha()
back_button_pushed = pygame.image.load('images\\back icon pushed.jpg')
back_button_pushed.set_colorkey((255, 255, 255))
pygame.draw.polygon(back_button_pushed, (255, 255, 255), [(164, 108), (148, 108), (164, 92)]) #get rid of the white triangle in the corner
back_button_pushed.convert_alpha()

other_buttons = []
for button in [pygame.image.load('images\\play icon.jpg'), pygame.image.load('images\\play icon pushed.jpg'), pygame.image.load('images\\instructions icon.jpg'), pygame.image.load('images\\instructions icon pushed.jpg')]:
    button.set_colorkey((255, 255, 255))
    pygame.draw.polygon(button, (255, 255, 255), [(327, 186), (311, 186), (327, 170)])
    pygame.draw.polygon(button, (255, 255, 255), [(0, 0), (16, 0), (0, 16)])
    pygame.draw.polygon(button, (255, 255, 255), [(327, 0), (311, 0), (327, 16)])
    pygame.draw.polygon(button, (255, 255, 255), [(0, 186), (16, 186), (0, 170)])
    button.convert_alpha()
    other_buttons.append(button)

#create the menu system
tree = Screen(pygame.image.load('images\\canyon background.jpg'),
              [Button(back_button, 'back', 82, 55, back_button_pushed), #exit button
               Button(other_buttons[0], #game
                      Game_Screen(
                          [Button(back_button, 'back', 82, 55, back_button_pushed)], []), 
                      width/2-300, height/2+200, other_buttons[1]), 
               Button(other_buttons[2], #instructions
                      Screen(pygame.image.load('images\\canyon background.jpg'),
                              [Button(back_button, 'back', 82, 55, back_button_pushed)], 
                              [],
                              [Image(pygame.image.load('images\\instructions.jpg'), width/2, height/2)]
                             ),
                      width/2+300, height/2+200, other_buttons[3]), 
               Button(pygame.Rect(0, 0, 0, 0), #score screen
                      Screen(pygame.image.load('images\\canyon background for score screen.jpg'),
                              [Button(back_button, 'back', 82, 55, back_button_pushed)], 
                              [Text('x', 200, None, (0, 0, 0), width/2, height/2-200)],
                              []
                             )
                      ),
               ],
              [], [Text('ENDLESS\nCANYON', 200, None, (190, 200, 210), 500, 270)]) #title


current_screen = tree
while True: #game loop
    events = pygame.event.get() #get events
    for event in events:
        if event.type == QUIT:
            pygame.quit()
            sys.exit()
        if event.type == KEYUP:
            if event.key == K_ESCAPE: #escape key triggers go_back()
                current_screen = go_back(tree, current_screen)
                if type(current_screen) == Game_Screen: pygame.mouse.set_visible(False)
                else: pygame.mouse.set_visible(True)
        if event.type == MOUSEBUTTONUP: #check for button clicks
            for button in current_screen.buttons:
                x, y = pygame.mouse.get_pos()
                if button.check_click(x, y):
                    button.pushed = False
                    current_screen = button.press(tree, current_screen)
                    if type(current_screen) == Game_Screen: pygame.mouse.set_visible(False)
                    else: pygame.mouse.set_visible(True)
        if event.type == MOUSEBUTTONDOWN:
            for button in current_screen.buttons:
                x, y = pygame.mouse.get_pos()
                if button.check_click(x, y):
                    button.pushed = True
    if type(current_screen) == Game_Screen: #run one iteration of the game loop if that is the current screen
        score = game.loop(window, events) #pass events into the game loop
        if score != None: #the game has finished
            game.setup(window) #reset the game
            tree.buttons[3].destination.texts[0].string = str(score)+'x' #update the score screen
            tree.buttons[3].destination.texts[0].update()
            current_screen = tree.buttons[3].destination #go to the score screen
            pygame.mouse.set_visible(True)
    current_screen.display()
    pygame.display.update()



Summative/game4.py
#ENDLESS CANYON Game Code
#Isaac Liao
#version 4
#23-05-2017
import pygame, sys, math, numpy, time, random
from pygame.locals import*
if __name__ == '__main__':
    __file__ = sys.path[0] + '         '
sys.path.append(__file__[:-9] + '\\modules\\mymath')
sys.path.append(__file__[:-9] + '\\modules\\graphics')
sys.path.append(__file__[:-9] + '\\modules\\sounds')
import mymath_v01 as mymath
import graphics_v5_2 as graphics
from math import*
from numpy import*

def player_height(pos): #returns the height of the ground at (x, y), better matches the display
    global world
    corner00 = world.terrain.loaded[world.terrain.distance][world.terrain.distance] #y position of corner00
    corner01 = world.terrain.loaded[world.terrain.distance][world.terrain.distance+1]
    corner10 = world.terrain.loaded[world.terrain.distance+1][world.terrain.distance]
    corner11 = world.terrain.loaded[world.terrain.distance+1][world.terrain.distance+1]
    pos00 = (pos/world.terrain.scale)%1 #xz position of the pos
    if pos00[0][0] > pos00[2][0]: #pos is on one triangle, use corner10
        base = corner10 + (corner00-corner10)*(1-pos00[0][0]) + (corner11-corner10)*(pos00[2][0])
        return world.terrain.scale * base
    else: #pos is on the other triangle, use corner01
        base = corner01 + (corner00-corner01)*(1-pos00[2][0]) + (corner11-corner01)*(pos00[0][0])
        return world.terrain.scale * base
def ground_height(pos): #returns the height of the ground at (x, y), more versatile
    global world
    return graphics.Mountains(pos[0][0]/world.terrain.scale, pos[2][0]/world.terrain.scale, world.terrain.func_state) * world.terrain.scale
def draw_aim(window, craft): #draw the target reticle
    w, h = window.get_width(), window.get_height()#draw the target reticle
    pygame.draw.line(window, (0, 0, 0), (int(w/2-10), int(h/2)), (int(w/2-30), int(h/2)), 2)
    pygame.draw.line(window, (0, 0, 0), (int(w/2+10), int(h/2)), (int(w/2+30), int(h/2)), 2)
    pygame.draw.line(window, (0, 0, 0), (int(w/2), int(h/2-10)), (int(w/2), int(h/2-30)), 2)
    pygame.draw.line(window, (0, 0, 0), (int(w/2), int(h/2+10)), (int(w/2), int(h/2+30)), 2)

#universal aircraft class, can support planes, helicopters, etc.
class Aircraft():
    def __init__(self, linear_drag, angular_drag):
        try: px, pz = player.craft.pos[0][0], player.craft.pos[2][0] #try to spawn at the player
        except NameError: pz, px = 0, 0
        x = random.uniform(-1000+px, 1000+px)
        z = random.uniform(-1000+pz, 1000+pz)
        y = ground_height(array([[x], [0], [z]])) + 100
        self.pos = array([[x], [y], [z]])
        self.theta = eye(3) #theta transforms from the plane's coordinate system to the world's.
        self.omega = zeros([3, 1]) #the angular velocity of the craft, this is in the plane's coordinate system
        self.v = zeros([3, 1]) #the linear velocity of the craft
        self.accelleration = zeros([3, 1]) #this is in the plane's coordinate system
        self.alpha = zeros([3, 1]) #the angular accelleration of the craft
        self.linear_drag = linear_drag
        self.angular_drag = angular_drag
        self.hp = 30
    def move(self): #update the aircraft
        self.v += self.theta.dot(self.accelleration) #accellerate
        self.omega += self.alpha #torque
        self.pos += self.v #move
        self.theta = self.theta.dot(mymath.rot_mat(self.omega, 'rotation vector')) #rotate
        self.v += array([[0], [-0.03], [0]]) #gravity
        self.v -= self.theta.dot(self.theta.T.dot(self.v)*self.linear_drag) #linear drag
        self.omega *= 1-self.angular_drag #angular drag
    def get_altitude(self): #returns the height relative to the ground
        return self.pos[1][0]-ground_height(self.pos)
    def control(self, F, tau, max_a, max_alpha): # F=force, tau=torque. This controls the aircraft.
        self.accelleration = F * max_a
        self.alpha = tau * max_alpha

#player class, similar to enemy class
class Player():
    def __init__(self, specs): #specs means specifications, a dictionary of the aircraft's drag and thrust
        global world
        self.camera = graphics.Camera()
        self.model = graphics.Model(world.ambient, world.lights, __file__[:-9]+'\\3D models\\jet', False)
        world.models.append(self.model)
        self.craft = Aircraft(specs['linear drag'], specs['angular drag'])
        self.max_a = specs['linear accelleration']
        self.max_alpha = specs['angular accelleration']
    def update(self): #update the player
        self.craft.move()
        offset = array([[-40], [6], [0]], dtype=float) #camera offset
        self.camera.pos = self.craft.pos + self.craft.theta.dot(offset) #move camera
        self.camera.theta = self.craft.theta #rotate camera
        self.model.pos = self.craft.pos #move model
        self.model.theta = self.craft.theta #rotate model
        if random.uniform() < 0.03*(30-self.craft.hp): #emit flames from damage
            global fragments
            fragments.append(Fragment('flames', self.craft.pos+(numpy.random.rand(3, 1)-0.5)*10, copy(self.craft.v / 2)))
    def shoot(self, mode='bullet'): #shoot either a missile or bullet
        if mode == 'bullet': #shoot a bullet
            global bullets
            bullet = Bullet(copy(self.craft.pos+140*mymath.unit(self.craft.v)), copy(self.craft.theta.dot(array([[100], [0], [0]], dtype=float))))
            bullets.append(bullet)
        elif mode == 'missile': #shoot a missile
            global missiles
            closest_aim = -1 #closest_aim tracks the distance from the target reticle to the enemy
            target=None
            for enemy in enemies+[player]: #find the aircraft which is the closest to the target reticle
                if enemy != self:
                    aim = self.craft.theta[:,0].dot(mymath.unit(enemy.craft.pos-self.craft.pos))
                    if closest_aim < aim and aim > 0.8:
                        target = enemy
                        closest_aim = aim
            missile = Missile(self.craft.pos+60*mymath.unit(self.craft.v), self.craft.v+5*mymath.unit(self.craft.v), target)
            missiles.append(missile)
    def crashtest(self): #test for ground collision and bullet/missile collision
        h = ground_height(self.craft.pos)
        altitude = self.craft.pos[1][0]-h
        if altitude < 0: #test for ground collision
            self.crash()
            return True
        for obj in bullets+missiles: #test for bullet/missile collision
            closest_point = mymath.closestpoint(self.craft.pos, obj.pos, obj.pos - obj.v)
            if mymath.norm(self.craft.pos - closest_point)<20:
                obj.pos = closest_point
                obj.hit(mymath.unit(obj.pos-self.craft.pos))
                if type(obj) == Bullet:
                    self.craft.hp -= 1
                    sounds['hit'+str(random.randint(1, 5))].play()
                elif type(obj) == Missile:
                    self.craft.hp -= 20
                    sounds['explosion'].play()
                if self.craft.hp <= 0:
                    self.crash()
                    return True
        return False
    def crash(self): #crash and explode
        global world, fragments
        world.models.remove(self.model)
        sounds['explosion'].play()
        for i in range(10):
            fragments.append(Fragment('plane frag', copy(self.craft.pos), (numpy.random.rand(3, 1)-0.5)*mymath.norm(self.craft.v)*10))

#enemy class, similar to player class
class Enemy():
    def __init__(self, specs): #specs is a dictionary of the aircraft's drag and thrust
        global world
        self.craft = Aircraft(specs['linear drag'], specs['angular drag'])
        self.model = graphics.Model(world.ambient, world.lights, __file__[:-9]+'\\3D models\\jet', False)
        world.models.append(self.model)
        self.max_a = specs['linear accelleration']
        self.max_alpha = specs['angular accelleration']
    def update(self): #update the enemy
        self.craft.move()
        self.model.pos = self.craft.pos #move model
        self.model.theta = self.craft.theta #rotate model
        for obj in bullets+missiles: #test for bullet/missile collision
            closest_point = mymath.closestpoint(self.craft.pos, obj.pos, obj.pos - obj.v)
            if mymath.norm(self.craft.pos-closest_point)<20:
                obj.pos = closest_point
                obj.hit(mymath.unit(closest_point-self.craft.pos))
                if type(obj)==Bullet: self.craft.hp -= 1
                elif type(obj)==Missile: self.craft.hp -= 20
                if self.craft.hp <= 0: self.crash()
        if random.uniform() < 0.03*(30-self.craft.hp): #emit flames from damage
            global fragments
            fragments.append(Fragment('flames', self.craft.pos+(numpy.random.rand(3, 1)-0.5)*10, copy(self.craft.v / 2)))
    def automate(self): #control the enemy aircraft
        F = array([[0.8], [0], [0]]) #continuous thrust
        altitude = self.craft.get_altitude()
        if altitude<100: avoid_crash = 150/altitude #avoid ground collision
        else: avoid_crash = 0
        stay_upright = expand_dims(cross(array([0, 1, 0]), self.model.theta[:,1]), 1) * 0.3 #stay upright
        attack = ((self.craft.theta.T.dot(cross(self.model.theta[:,0], mymath.unit(self.craft.pos - player.craft.pos).T).T)>0)-0.5).astype(float) * 0.5 #attack player
        tau = expand_dims(cross(array([0, avoid_crash, 0]), self.model.theta[:,0]), 1)
        tau += stay_upright
        tau += attack
        self.craft.control(F, tau, self.max_a, self.max_alpha) #control the aircraft
        if self.model.theta[:,0].T.dot(mymath.unit(player.craft.pos - self.craft.pos)) > 0.9: #shoot if player is within 10 degrees of target reticle
            global bullets
            self.shoot()
    def crashtest(self): #test for ground collision
        h = ground_height(self.craft.pos)
        altitude = self.craft.pos[1][0]-h
        if altitude < 0:
            self.crash()
            return
    def crash(self): #crash and explode
        global world, fragments, enemies, kills
        kills += 1
        try:
            enemies.remove(self)
            world.models.remove(self.model)
        except ValueError:
            pass
        sounds['explosion'].play()
        for i in range(10):
            fragments.append(Fragment('plane frag', copy(self.craft.pos), (numpy.random.rand(3, 1)-0.5)*mymath.norm(self.craft.v)*10))
    def shoot(self, mode='bullet'): #shoot
        if mode == 'bullet':
            global bullets
            bullet = Bullet(self.craft.pos+140*mymath.unit(self.craft.v), self.craft.theta.dot(array([[100], [0], [0]], dtype=float)))
            bullets.append(bullet)
        elif mode == 'missile': #enemy AI will not shoot missiles
            global missiles
            closest_aim = -1 #equivalent to the distance from the target reticle to the enemy
            target=None
            for enemy in enemies+[player]:
                if enemy != self:
                    aim = self.craft.theta[0].dot(mymath.unit(enemy.craft.pos-self.craft.pos))
                    if closest_aim < aim and aim > 0.8:
                        target = enemy
                        closest_aim = aim
            missile = Missile(self.pos+60*mymath.unit(self.craft.v), self.craft.v+5*mymath.unit(self.craft.v), target)
            missiles.append(missile)
class Bullet(): #bullet class
    def __init__(self, pos, v):
        global world
        self.model = graphics.Model(world.ambient, world.lights, __file__[:-9]+'\\3D models\\bullet', False)
        world.models.append(self.model)
        self.pos = pos
        self.v = v
    def move(self): #move bullet
        global world
        self.pos += self.v
        self.model.pos = self.pos
        if mymath.norm(player.craft.pos - self.pos) > world.terrain.scale * world.terrain.distance: #despawn if too far
            try:
                bullets.remove(self)
                world.models.remove(self.model)
            except ValueError:
                pass
        elif self.get_altitude()<0: #hit ground
            v1 = ground_height(self.pos)
            v2 = ground_height(self.pos + array([[1], [0], [0]]))
            v3 = ground_height(self.pos + array([[0], [0], [1]]))
            normal = -expand_dims(mymath.unit(cross(array([1, v2-v1, 0]), array([0, v3-v1, 1]))), 1)
            self.hit_ground(normal)
    def get_altitude(self): #get altitude
        return self.pos[1][0]-ground_height(self.pos)
    def hit_ground(self, normal): #hit the ground, do not generate fragments, instead make dirt clumps.
        global world, bullets, fragments
        try:
            bullets.remove(self)
            world.models.remove(self.model)
        except ValueError:
            pass
        for i in range(4):
            fragments.append(Fragment('dirt clump', copy(self.pos), (self.v - 2*normal*normal.T.dot(self.v))*0.3+ (numpy.random.rand(3, 1)-0.5)*mymath.norm(self.v)*0.1))
    def hit(self, normal): #hit something other than the ground, generate fragments
        global world, bullets, fragments
        try:
            bullets.remove(self)
            world.models.remove(self.model)
        except ValueError:
            pass

        #create fragments upon collision
        for i in range(4):
            fragments.append(Fragment('bullet', copy(self.pos), self.v - 2*normal*normal.T.dot(self.v) + (numpy.random.rand(3, 1)-0.5)*mymath.norm(self.v)*0.5))

class Missile(): #missile class
    def __init__(self, pos, v, target):
        global world
        self.model = graphics.Model(world.ambient, world.lights, __file__[:-9]+'\\3D models\\missile', False)
        world.models.append(self.model)
        self.pos = pos
        self.v = v
        self.target = target
    def move(self): #move the missile
        global world
        if self.target!=None:self.v += mymath.unit(self.target.craft.pos - self.pos)*5 #thrust
        else: self.v += mymath.unit(self.v)*5
        self.v *= 0.9 #drag
        self.pos += self.v
        self.model.pos = self.pos
        side = cross(mymath.unit(self.v).T, array([[0, 1, 0]])).T #a column vector for the side of the missile
        self.model.theta = hstack([mymath.unit(self.v), -side, cross(mymath.unit(self.v).T, side.T).T ]) #rotate model
        if self.get_altitude()<0: #hit ground
            v1 = ground_height(self.pos)
            v2 = ground_height(self.pos + array([[1], [0], [0]]))
            v3 = ground_height(self.pos + array([[0], [0], [1]]))
            normal = -expand_dims(mymath.unit(cross(array([1, v2-v1, 0]), array([0, v3-v1, 1]))), 1)
            self.hit(normal)
        if self.target == None and mymath.norm(player.craft.pos - self.pos) > world.terrain.scale * world.terrain.distance: #despawn if too far
            global missiles
            try:
                missiles.remove(self)
                world.models.remove(self.model)
            except ValueError:
                pass
    def get_altitude(self): #get altitude
        return self.pos[1][0]-ground_height(self.pos)
    def hit(self, normal): #collide and explode
        global world, missiles, fragments
        try:
            missiles.remove(self)
            world.models.remove(self.model)
        except ValueError:
            pass

        sounds['explosion'].play()

        #create flames upon collision
        for i in range(10):
            fragments.append(Fragment('flames', copy(self.pos), self.v - 2*normal*normal.T.dot(self.v) + (numpy.random.rand(3, 1)-0.5)*mymath.norm(self.v)*0.5))

class Fragment(): #fragment class for display effects
    def __init__(self, filename, pos, v):
        global world
        self.model = graphics.Model(world.ambient, world.lights, __file__[:-9]+'\\3D models\\' + filename, False)
        world.models.append(self.model)
        self.pos = pos
        self.v = v
        self.time = 5 #fragments should disappear after 5 frames
    def move(self): #update/move
        self.time -= 1
        self.pos += self.v
        self.model.pos = self.pos
        if self.time < 0:
            global world, fragments
            fragments.remove(self)
            world.models.remove(self.model)

def setup(window): #this runs once, at the beginning of a game
    global jet, helicopter, tester #this game only uses the jet
    jet = {'linear drag':array([[0.01], [0.7], [0.3]]), 'angular drag':array([[0.15], [0.4], [0.6]]), 'linear accelleration':array([[0.2], [0.01], [0.04]]), 'angular accelleration':array([[0.015], [0.02], [0.1]])}
    helicopter = {'linear drag':array([[0.015], [0.04], [0.06]]), 'angular drag':array([[0.4], [0.7], [0.7]]), 'linear accelleration':array([[0], [0.1], [0.01]]), 'angular accelleration':array([[0.007], [0.1], [0.05]])}
    tester = {'linear drag':array([[0.01], [0.01], [0.01]]), 'angular drag':array([[0.4], [0.4], [0.4]]), 'linear accelleration':array([[0.2], [0.2], [0.2]]), 'angular accelleration':array([[0.007], [0.1], [0.1]])}

    global world #the world contains all 3D graphics imformation except for the sky's colour.
    world = graphics.WorldModel()
    world.ambient = (0.1, 0.1, 0.4) #blue ambient light
    world.lights.append(graphics.Light(direction=mymath.unit(array([[3], [-1], [1]])), colour=(1, 1, 1))) #white angled sunlight
    terrain_scale = 300 #set up the terrain
    terrain_state = {'detail':4, 'max_dist':8, 'smoothness':2, 'max_height':30, 'seed':random.randint(1, 1<<15)}
    world.terrain = graphics.Terrain(graphics.Mountains, terrain_state, (0.9, 0.4, 0.2), terrain_scale, 6, world.ambient, world.lights)

    global background_colour, player
    background_colour = (120, 150, 255) #blue sky
    player = Player(jet) #create player

    global enemies, bullets, fragments, missiles #create lists containing other objects
    enemies = []
    bullets = []
    fragments = []
    missiles = []

    global F, tau, mousehold
    F = zeros([3, 1]) #supposed to go from -1 to 1
    tau = zeros([3, 1]) #supposed to go from -1 to 1
    mousehold = True #hold the mouse in the center

    global kills, framecounter, shooting, missiles_left
    kills = 0 #count the kills
    framecounter = 0
    shooting = False
    missiles_left = 5 #the player is given 5 missiles

    global sounds #load all sounds
    sounds = {'shooting':pygame.mixer.Sound(__file__[:-9]+'\\sounds\\shooting.wav'),
              'explosion':pygame.mixer.Sound(__file__[:-9]+'\\sounds\\explosion.wav'),
              'hit1':pygame.mixer.Sound(__file__[:-9]+'\\sounds\\hit1.wav'),
              'hit2':pygame.mixer.Sound(__file__[:-9]+'\\sounds\\hit2.wav'),
              'hit3':pygame.mixer.Sound(__file__[:-9]+'\\sounds\\hit3.wav'),
              'hit4':pygame.mixer.Sound(__file__[:-9]+'\\sounds\\hit4.wav')}

def loop(window, events): #this runs as long as the game is running
    global world, player, enemies, bullets, fragments, missiles, F, tau, mousehold, framecounter, shooting, kills, sounds, missiles_left
    width = window.get_width()
    height = window.get_height()
    scroll = 0
    for event in events:
        if event.type == QUIT:
            pygame.quit()
            sys.exit()
        if event.type == KEYDOWN: #record key pressing
            if event.key == K_w:
                F[0][0] = 1 #key pressing is linked to craft linear accelleration
            if event.key == K_s:
                F[0][0] = -2
            if event.key == K_a:
                F[2][0] = -1
            if event.key == K_d:
                F[2][0] = 1
            if event.key == K_LSHIFT:
                F[1][0] = -1
            if event.key == K_SPACE:
                F[1][0] = 1
            if event.key == K_LCTRL:
                mousehold = False
                pygame.mouse.set_visible(True)
        if event.type == KEYUP:
            if event.key == K_w or event.key == K_s:
                F[0][0] = 0
            if event.key == K_a or event.key == K_d:
                F[2][0] = 0
            if event.key == K_LSHIFT or event.key == K_SPACE:
                F[1][0] = 0
            if event.key == K_LCTRL:
                mousehold = True
                pygame.mouse.set_visible(False)
        if event.type == MOUSEBUTTONDOWN:
            if event.button == 4: #scroll position is linked to craft angular velocity
                tau[0][0] += 1
            if event.button == 5:
                tau[0][0] -= 1
            if event.button == 1: #shoot bullet
                shooting = True
            if event.button == 3 and missiles_left > 0: #shoot missile
                missiles_left -= 1
                player.shoot('missile')
        if event.type == MOUSEBUTTONUP: #stop shooting bullets
            if event.button == 1:
                shooting = False
    if mousehold: #mouse position is linked to craft angular velocity
        pygame.mouse.set_pos((width/2, height/2))
        (x, y) = pygame.mouse.get_pos()
        tau[1][0] += (y-width/2)/750
        tau[2][0] -= (x-height/2)/750

    if framecounter >= 300 or len(enemies) == 0: #add an enemy every 300 frames
        enemies.append(Enemy(jet))
        framecounter = 1

    for obj in bullets+missiles+fragments: #update objects
        obj.move()
    for enemy in enemies: #update enemies
        enemy.update()
        enemy.automate()
        enemy.crashtest()
    player.craft.control(F, tau, player.max_a, player.max_alpha) #control the player
    if shooting:
        player.shoot('bullet')
        channel = pygame.mixer.find_channel()
        if channel != None:
            channel.play(sounds['shooting'])
            channel.fadeout(300)
        
    player.update() #update the player
    if player.crashtest(): #show 30 more frames on death
        sounds['explosion'].play()
        for frame in range(30):
            for obj in bullets+fragments+missiles:
                obj.move()
            for enemy in enemies:
                enemy.update()
                enemy.automate()
                enemy.crashtest()
            window.fill(background_colour)
            world.display(window, player.camera)
            pygame.display.update()
        return kills #finish game

    world.terrain.generate(vstack([player.craft.pos[0], player.craft.pos[2]]), world.ambient, world.lights) #generate terrain
    window.fill(background_colour)
    world.display(window, player.camera) #display the world
    draw_aim(window, player.craft) #draw the target reticle
    framecounter += 1
    return None

